corehq/apps/user_importer/validation.py

```diff
@@ -139,6 +142,39 @@ def validate_spec(self, spec):
             return self.error_message
 
 
+class TableauRoleValidator(ImportValidator):
+    _error_message = _("Invalid tableau role: '{}'. Please choose one of the following: {}")
+
+    def __init__(self, domain):
+        super().__init__(domain)
+        self.valid_role_options = [e.value for e in TableauUser.Roles]
+
+    def validate_spec(self, spec):
+        tableau_role = spec.get('tableau_role')
+        if tableau_role is not None and tableau_role not in self.valid_role_options:
+            return self._error_message.format(tableau_role, ', '.join(self.valid_role_options))
+
+
+class TableauGroupsValidator(ImportValidator):
+    _error_message = _("These groups, {}, are not valid for this domain. Please choose from the following: {}")
+
+    def __init__(self, domain):
+        super().__init__(domain)
+        self.allowed_groups_for_domain = None
+
+    def validate_spec(self, spec):
+        tableau_groups = spec.get('tableau_groups') or []
+        if tableau_groups:
+            tableau_groups = tableau_groups.split(',')
+            self.allowed_groups_for_domain = get_allowed_tableau_groups_for_domain(self.domain)
+        invalid_groups = []
+        for group in tableau_groups:
+            if group not in self.allowed_groups_for_domain:
```

Comment:
This isn't currently a problem, but should `self.allowed_groups_for_domain` be initialized as an empty list since `None` could error here.

nit: is there a reason to keep this as an instance attribute?  But how it is now, it'll hit the db for each row (since `validate_spec` is called for each row) that has `tableau_groups` defined. I believe it'll only need to hit the db once if you called `get_allowed_tableau_groups_for_domain` in `init` so I think that's preferable.

Reply:
> should self.allowed_groups_for_domain be initialized as an empty list since None could error here.

I'd say something like this would be the way to go:
`self.allowed_groups_for_domain = get_allowed_tableau_groups_for_domain(self.domain) or []`
if we initialize self.allowed_groups_for_domain to an empty list it could be updated to be `None` by `get_allowed_tableau_groups_for_domain`

> it'll only need to hit the db once if you called get_allowed_tableau_groups_for_domain in init

This is a good point. In a previous commit I moved `get_allowed_tableau_groups_for_domain` out of `init` and into `validate_spec` because I was hitting an error when there was no col for tableau groups, but I think I can pass `all_specs` into this validator similar to how it is in `DuplicateValidator`. That way I can check for the groups col in the spec and set `self.tableau_groups` and `self.allowed_groups_for_domain` in the `init`.